llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp

```diff
@@ -4237,6 +4242,70 @@ getFalkorUnrollingPreferences(Loop *L, ScalarEvolution &SE,
   }
 }
 
+static bool shouldUnrollLoopWithInstruction(Instruction &I,
+                                            AArch64TTIImpl &TTI) {
+  // Don't unroll vectorised loop.
+  if (I.getType()->isVectorTy())
+    return false;
+
+  if (isa<CallInst>(I) || isa<InvokeInst>(I)) {
+    if (const Function *F = cast<CallBase>(I).getCalledFunction())
+      if (!TTI.isLoweredToCall(F))
+        return true;
+    return false;
+  }
+
+  return true;
+}
+
+static bool shouldUnrollSmallMultiExitLoop(Loop *L, ScalarEvolution &SE,
```

Comment:
This looks like a subset of `getAppleRuntimeUnrollPreferences`, can we unify it?

Reply:
Sure, I could do that although it's actually a superset because in `getAppleRuntimeUnrollPreferences` it seems to disallow loops with multiple exit blocks, i.e. both exiting blocks must have the same destination.

Reply:
Hmm, sorry, maybe that's not right. I'm just looking at this code:

```
  // Small search loops with multiple exits can be highly beneficial to unroll.
  if (!L->getExitBlock()) {
```

Perhaps I've misunderstood what this is doing?

Reply:
I thought at first this is saying if the loop doesn't have a single exit block then unroll, however looking at the implementation of getExitBlock a bit more I can see that it doesn't permit repeats. So I guess that means it should always return false for any early exit loop.

Reply:
The code block should handle small multi exit loops as advertised (tests are in `llvm/test/Transforms/LoopUnroll/AArch64/apple-unrolling-multi-exit.ll`). `getExitBlock()` should return the single exit block if the loop has a single exit & exiting block. 

If there are multiple exiting blocks to the same exit, the exit block will be included multiple times in getExitBlocks(); This is different from `getUniqueExitBlock`, which would return the common exit block, even if there are multiple exiting blocks branching to it.

Reply:
Hi, I'm not sure exactly what you meant by 'unify' and how far I should attempt to commonise code. I've tried my best - see what you think! The main difference is that now getAppleRuntimeUnrollPreferences bails out on multi-exit loops and defers to common code. The common code bails out for slightly more cases, for example if an intrinsic gets lowered to a call, if the loop already contains a vector type, or if a cost cannot be calculated. I assume that these are sensible things to do for all targets anyway. One side-effect of these changes is that it's now possible to override the default unroll factor of 4 - I've added a test to Transforms/LoopUnroll/AArch64/apple-unrolling-multi-exit.ll for forcing unroll=2.